File: 2586-longest-square-streak-in-an-array/2586-longest-square-streak-in-an-array.java
class Solution {
    public int longestSquareStreak(int[] nums) {
        Arrays.sort(nums);
        Map<Integer, Integer> dp = new HashMap<>();
        int result = 0;

        for (int num : nums) {
            if (dp.containsKey((int) Math.sqrt(num)) && Math.sqrt(num) == (int) Math.sqrt(num)) {
                dp.put(num, dp.get((int) Math.sqrt(num)) + 1);
            } else {
                dp.put(num, 1);
            }
            result = Math.max(result, dp.get(num));
        }

        return result == 1 ? -1 : result;
    }
}

File: 2586-longest-square-streak-in-an-array/2586-longest-square-streak-in-an-array.cpp
class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        unordered_set<int> st(nums.begin(), nums.end());
        int maxi=0;
        for(int i=0;i<nums.size();i++)
        {
            int count=0;
            long long curr=(long long)nums[i];
            while(st.find(curr)!=st.end())
            {
                count++;
                curr=(long long)curr*curr;
                if(curr>INT_MAX)
                    break;
            }
            maxi=max(maxi, count);
        }
        if(maxi<2)
            return -1;
        return maxi;
    }
};